src/main/scala/viper/gobra/frontend/ParseTreeTranslator.scala

```diff
@@ -0,0 +1,2271 @@
+    }
+  }
+
+  /** Helper Function for optional Nodes
+    *
+    * @param ctx a context that might be null (signified with ? in the grammar)
+    * @tparam P The PNode type
+    * @return a positioned Option of a positioned PNode
+    */
+  def visitNodeOrNone[P <: PNode](ctx : ParserRuleContext) : Option[P] = {
+    if (ctx != null) {
+      Some(visitNode(ctx)).pos()
+    } else None
+  }
+
+  /** Helper Function for Nodes with a default
+    *
+    * @param ctx a context that might be null (signified with ? in the grammar)
+    * @tparam P The PNode type
+    * @return a positioned Option of a positioned PNode
+    */
+  def visitNodeOrElse[P <: PNode](ctx : ParserRuleContext, default : P) : P = {
+    visitNodeOrNone[P](ctx) match {
+      case Some(value) => value
+      case None => default
+    }
+  }
+
+  def visitListNode[P <: PNode](ctx : java.util.List[_ <: ParserRuleContext]) : Vector[P] = {
+    ctx.asScala.view.map(visitNode[P]).toVector match {
+      case e@Vector() => e
+      case v => v.range(v.head, v.last)
+    }
+  }
+
+  def visitNodeIf[P <: PNode, C <: ParserRuleContext](ctx : java.util.List[C], cond : (C => Boolean)) : Vector[P] = {
+    ctx.asScala.toVector.collect {
+      case c if cond(c) => visitNode(c)
+    }.asInstanceOf[Vector[P]]
+  }
+  //endregion
+
+  //region Error reporting and positioning
+
+  private def fail(cause : NamedPositionable, msg : String = "") = {
+    throw TranslationFailure(cause, msg)
+  }
+
+  private def unexpected(ctx: ParserRuleContext) = {
+    violation(s"could not translate '${ctx.getText}', got unexpected production '${visitChildren(ctx)}'")
+  }
+
+  private def warn(cause: NamedPositionable, msg : String): Unit = {
+    warnings.append(TranslationWarning(cause, msg))
+  }
+
+  implicit class PositionableParserRuleContext[P <: ParserRuleContext](ctx: P) extends NamedPositionable {
+    val startPos : Position = ctx.getStart.startPos
+    val endPos : Position = if (ctx.getStop == null) ctx.getStart.startPos else ctx.getStop.endPos
+    val name : String = GobraParser.ruleNames.array(ctx.getRuleIndex)
+    val text : String = ctx.getText
+  }
+
+  implicit class PositionableTerminalNode[T <: TerminalNode](term: T) extends NamedPositionable {
+    private val tok = term.getSymbol
+    val startPos : Position = tok.startPos
+    val endPos : Position = tok.endPos
+    val name : String = GobraParser.VOCABULARY.getDisplayName(tok.getType)
+    val text : String = term.getText
+  }
+
+  implicit class PositionableToken[T <: Token](tok: T) extends NamedPositionable {
+    val startPos : Position = Position(tok.getLine, tok.getCharPositionInLine + 1, source)
+    val endPos : Position = Position(tok.getLine, tok.getCharPositionInLine + 1 +(tok.getStopIndex -tok.getStartIndex), source)
+    val name : String = GobraParser.VOCABULARY.getDisplayName(tok.getType)
+    val text : String = tok.getText
+  }
+
+  implicit class PositionedPNodeOption[O <: Some[PNode]](some: O) {
```

Comment:
I'm a bit skeptical seeing three highly similar implicit classes. Couldn't we make them more generic such that we only need one of them?

Reply:
I've somewhat generalized it now.